pkg/maintainer/spv/spv.go

```diff
@@ -57,10 +105,348 @@ func (sm *spvMaintainer) startControlLoop(ctx context.Context) {
-	<-ctx.Done()
+		// TODO: Add proving of other type of SPV transactions: redemption
+		// transactions, moving funds transaction, etc.
+
+		select {
+		case <-time.After(sm.config.IdleBackOffTime):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (sm *spvMaintainer) proveDepositSweepTransactions() error {
+	depositSweepTransactions, err := sm.getUnprovenDepositSweepTransactions()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get unproven deposit sweep transactions: [%v]",
+			err,
+		)
+	}
+
+	txProofDifficultyFactor, err := sm.spvChain.TxProofDifficultyFactor()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get transaction proof difficulty factor: [%v]",
+			err,
+		)
+	}
+
+	// TODO: Handle a situation in which the block headers in the proof span
+	//       multiple Bitcoin difficulty epochs.
+	requiredConfirmations := uint(txProofDifficultyFactor.Uint64())
+
+	for _, transaction := range depositSweepTransactions {
+		accumulatedConfirmations, err := sm.btcChain.GetTransactionConfirmations(
+			transaction.Hash(),
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to get transaction confirmations: [%v]",
+				err,
+			)
+		}
+
+		if accumulatedConfirmations < requiredConfirmations {
+			// Skip the transaction as it has not accumulated enough
+			// confirmations. It will be proven later.
+			continue
+		}
+
+		_, proof, err := bitcoin.AssembleSpvProof(
+			transaction.Hash(),
+			requiredConfirmations,
+			sm.btcChain,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to assemble SPV proof: [%v]", err)
+		}
+
+		// Deserialize the first block header and verify its difficulty.
+		rawBlockHeader := [bitcoin.BlockHeaderByteLength]byte{}
+		copy(
+			rawBlockHeader[:],
+			proof.BitcoinHeaders[:bitcoin.BlockHeaderByteLength],
+		)
+		firstBlockHeader := bitcoin.BlockHeader{}
+		firstBlockHeader.Deserialize(rawBlockHeader)
+		firstBlockHeaderDifficulty := firstBlockHeader.Difficulty()
+
+		currentDifficulty, previousDifficulty, err :=
+			sm.btcDiffChain.GetCurrentAndPrevEpochDifficulty()
+		if err != nil {
+			return fmt.Errorf(
+				"failed to get Bitcoin epoch difficulties: [%v]",
+				err,
+			)
+		}
+
+		if firstBlockHeaderDifficulty.Cmp(currentDifficulty) != 0 &&
+			firstBlockHeaderDifficulty.Cmp(previousDifficulty) != 0 {
+			// Skip the transaction as the difficulty of the first block
+			// header in the proof does not match the current or the
+			// previous difficulty seen on-chain. The reason for this is
+			// most likely that the proof consists of block headers from the
+			// beginning of a new Bitcoin difficulty epoch and the epoch is
+			// not proven on-chain yet. In that case the transaction will be
+			// proven in the future. The other reason could be that the
+			// transaction is older than the last two Bitcoin difficulty
+			// epoch. In that case the transaction will soon leave the
+			// sliding window of recent transactions.
+			continue
```

Comment:
I think we should log a warning in this case. The second case is pretty dangerous. 

We should also make sure those warnings are included in our monitoring rules cc @lukasz-zimnoch.

Reply:
Done in 132ac7d666c8964b47a71255b3c50334f240fe22.